### DISCUSSION TITLE
Cannot import other functions with Python API

### DISCUSSION THREAD

**Discussion by ninest** [vercel/vercel#4717] [Help]
After seeing this [comment](https://github.com/vercel/vercel/issues/3729#issuecomment-582155572), I learned that to import a function in your Python API, you have to put an underscore before the file name.

I tried this method but it doesn't seem to work. I have `_utils/common.py` inside my `api/` directory.

My `index.py` looks like this:

```python
from flask import Flask, Response
from _utils.common import VAR
app = Flask(__name__)

print(VAR)

...
```

When I try to visit any API route, I get this error:
```
 File "./api/index.py", line 2, in <module>
    from _utils.common import VAR
ModuleNotFoundError: No module named '_utils'
```
Everything works fine when I run the code with `python3 api/index.py`, but using `vercel dev` causes this error. It comes as a 502: bad gateway error.

Is there an official example I can refer to?

**Comment by kittinan** [comment]
@ninest Did you found solution to import  file?

**Reply by ninest** [reply]
Sorry, no. I haven't tried anything since I posted this

**Comment by kittinan** [comment]
@ninest I found solution.

Original project structure (Not working)

```
ROOT PROJECT
  |
  |
 API
  |
  | -- util
  |      | -- spotify.py
  |
  | -- index.py

```
```
# index.py

from util import spotify
```

I moved `/api/util` folder to root path project `/util`. index.py can import correctly when running on Vercel serverless function

**Reply by ernesto** [reply]
That´s no make sense... Why with 
```
"functions": {
    "api/index.py": {
      "includeFiles": "api/_utils.py"
    }
}
```
Doesnt works?

**Reply by GoulartNogueira** [reply]
@ernesto 
I believe "includeFiles" means files used to generate APIs.
It's not about import.

**Comment by shimahi** [comment]
@ninest 

If you set `vercel.json` on project root directory, try to write import phrase to refer absolute path from `vercel.json`, or relative path from `index.py` .
```
from api.util.spotify import 〜〜〜
```
or
```
from .util.spotify import 〜〜〜
```



cf. vercel.json is like this.
```
{
  "functions": {
    "api/index.py": {
      "includeFiles": "_utils/**"
    }
  },
  "routes": [{ "src": "/", "dest": "api/index.py" }]
}
```

**Comment by FVitor7** [comment]
Solution?

**Reply by ninest** [reply]
If I'm not wrong, you can import the python functions as your normally would in a python program, no extra steps needed. I'll try making an example

**Reply by FVitor7** [reply]
I would appreciate your example, I'm trying everything and I can't import other .py files

**Reply by ninest** [reply]
Actually, I do have an example here 

https://github.com/ninest/ippt

It's a fitness test score calculator. The score calculation is in the `calculator` folder. In the `api` folder, I'm using flask, and importing functions from `calculator`. See if it helps solve your problem

**Reply by FVitor7** [reply]
It worked!!! thank you so much

**Comment by pado31** [comment]
I have a similar problem. My project is structured as follows:
```
ROOT PROJECT
  |
  |
 src
  |
  | -- util
  |      | -- __init__.py
  |      | -- service.py
  |
  | -- server.py
  | -- vercel.json
  | -- requirements.txt
```

Inside server.py I have the following code:
```
from flask import Flask
from util.service import Updater

app = Flask(__name__)

@app.route("/api/v1/update/")
def getUpdate():
    return Updater.update_values()
```

And my vercel.json:
```
{
    "version": 2,
    "builds": [
        {
            "src": "src/server.py",
            "use": "@vercel/python"
        }
    ],
    "routes": [
        {
            "src": "/",
            "dest": "src/server.py"
        }
    ]
}
```

Vercel gives me the following error:
`[ERROR] Runtime.ImportModuleError: Unable to import module 'vc__handler__python': No module named 'util'`

Locally my project runs correctly. Moreover, if in server.py I don't use any import and I simply write `return "Hello, world!"`, the api works so I suppose that the configuration is right.

Any solution? Thank you!

**Reply by warmachine028** [reply]
Yes I am also having the same error

**Reply by tanw89** [reply]
Same issue too. Anyone found a workaround?

**Comment by deepakstark01** [comment]
```
{
  "functions": {
    "api/index.py": {
      "includeFiles": [
        "_utils/**",
        "other_functions/**"
      ]
    }
  },
  "routes": [{ "src": "/", "dest": "api/index.py" }]
}
```

**Comment by deepakstark01** [comment]
```
{
  "routes": [
    {
      "src": "/api/(.*)",
      "dest": "api/index.py"
    }
  ]
}

```

**Comment by Chaoyingz** [comment]
You can add the environment variable `PYTHONPATH` to help the Python interpreter locate this module.
Add this to vercel.json:

```json
"env": {
    "PYTHONPATH": "api/"
}
```

**Reply by Upendra-Thunuguntla** [reply]
Either do this or just add same in the Env Variable

**Reply by Marvy101** [reply]
```
"env": {
    "PYTHONPATH": "api/"
}
```

Doing this worked for me.

**Reply by antheco** [reply]
this worked for me

**Comment by FrancescoSaverioZuppichini** [comment]
can somehow share a working solution?

**Reply by deepakstark01** [reply]
yes

**Comment by Alfredoeb9** [comment]
After trail and error, this is what worked for me, using python in the backend with flask

folder structure:

```
ROOT PROJECT
  |
  |
 src
  |
  | -- api
  |      | -- index.py
  |      | -- models
  |            | -- models.py
  |
  | -- vercel.json
  | -- requirements.txt
```

vercel.json

```
{
  "rewrites": [
    {
      "source": "/(.*)",
      "destination": "/api/index"
    }
  ]
}
```

Within my api/index.py I am now able to import models.py but I have to specify the api folder when importing as so:

api/index.py: 
```
from api.models.models import <func>
```

Saved these changes and deployed in vercel and I did not get a 404 or 500. Hope these configs help someone out.
